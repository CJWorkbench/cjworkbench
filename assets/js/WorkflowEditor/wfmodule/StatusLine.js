// ---- StatusLine ----

// Display error message, if any
// BUG - Tying this to Props will ensure that error message stays displayed, even after resolution
import React, { useCallback, useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import QuickFix, { QuickFixPropTypes } from './QuickFix'

const StatusLine = React.memo(function StatusLine ({ status, errors, applyQuickFix, isReadOnly }) {
  const [clickedAnyQuickFix, setClickedQuickFix] = useState(false)
  const doApplyQuickFix = useCallback((...args) => {
    setClickedQuickFix(true)
    applyQuickFix(...args)
  })

  // after props change (remember: we're in React.memo), assume the quick fix
  // suggestions are not-yet-clicked.
  useEffect(() => setClickedQuickFix(false), [status, errors])

  if (!errors.length) return null

  return (
    <>
      {errors.map(({ message, quickFixes }, j) => (
        <div className='wf-module-error-msg' key={j}>
          <p>{message}</p>
          {quickFixes && quickFixes.length && !isReadOnly ? (
            <ul className='quick-fixes'>
              {quickFixes.map((quickFix, i) => (
                <li key={i}>
                  <QuickFix
                    disabled={clickedAnyQuickFix}
                    applyQuickFix={doApplyQuickFix}
                    {...quickFix}
                  />
                </li>
              ))}
            </ul>
          ) : null}
        </div>
      ))}
    </>
  )
})
StatusLine.propTypes = {
  status: PropTypes.oneOf(['ok', 'busy', 'error', 'unreachable']).isRequired,
  isReadOnly: PropTypes.bool.isRequired, // if true, cannot apply quick fixes
  errors: PropTypes.arrayOf(PropTypes.shape({ message: PropTypes.string.isRequired, quickFixes: PropTypes.arrayOf(PropTypes.shape(QuickFixPropTypes)) })), // may be empty
  applyQuickFix: PropTypes.func.isRequired // func(action, args) => undefined
}

export default StatusLine
