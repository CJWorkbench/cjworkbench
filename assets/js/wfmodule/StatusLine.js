// ---- StatusLine ----

// Display error message, if any
// BUG - Tying this to Props will ensure that error message stays displayed, even after resolution
import React from "react";
import PropTypes from "prop-types";

export default class StatusLine extends React.Component {
  render() {
    if (this.props.status == 'error') {
      return <div className='wf-module-error-msg mb-3'>{this.props.error_msg}</div>
    // } else if (this.props.status == 'busy') {
    //   return <div className='wf-module-error-msg mb-3'>Working...</div>
    } else {
      return false
    }
  }
}

StatusLine.PropTypes = {
  status: PropTypes.string,
  error_msg: PropTypes.string
}