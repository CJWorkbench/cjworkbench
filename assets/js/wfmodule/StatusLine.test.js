/* globals describe, it, expect */
import React from 'react'
import StatusLine from './StatusLine'
import { shallow } from 'enzyme'

describe('Status line', () => {
  it('Renders an error message', () => {
    let wrapper = shallow(
      <StatusLine
        status='error'
        error_msg="There's an error"
      />
    )
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('div').first().text()).toEqual("There's an error")
  })

  it('Renders nothing for other statuses', () => {
    let wrapper = shallow(
      <StatusLine
        status='ready'
        error_msg='This should never happen'
      />
    )
    expect(wrapper.find('div').length).toEqual(0)
  })
})
