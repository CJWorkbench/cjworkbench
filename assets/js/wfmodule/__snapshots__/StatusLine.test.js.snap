// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Status line Renders an error message 1`] = `
<div
  className="wf-module-error-msg mb-3"
>
  There's an error
</div>
`;
